### PR TITLE
47117332_Tong_Qiu_Siamese

### PR DESCRIPTION
**Pull request by Amelia-Tong**
pull request made late with an extension.

This repo should be run properly with correct environment setting.

Problems not solved:
- Did not meet the requirement of 80% test accuracy.
- The highest test accuracy achieved is 77.63%. This might be due to overfitting the validation set.
- More data augmentation techniques should be tried, but there wasn't enough time for further experiments.

Documents added:
- README.md: explained the algorithm, data, usage, and results.
- utils.py: training configurations, such as data pths, log pth, model pth, and device.
- dataset.py: self-defined dataset classes for contrastive learning,  triplet learning, and the classification task.
- modules.py: includes model components.
- train.py: source code for training, validating, testing and saving models.
- predict.py: code to get test accuracies, image-level prediction, and patient-level predictions.
- readme_imgs: images used in the readme file.
- sh_files: sh files for model training.

### CODE REVIEW

**Comment by nathasha-naranpanawa** [comment]
**This is an initial inspection, no action is required at this point**

Difficulty: Hard

Readme: Very good
- Clear flow of information
- Algortihm and usage explained well
- Relevant plots for loss and accuracies provided
- no 2D manifolds are presented

Commit messages: OK

Code: 
- tried both siamese and triplet networks
- good design
- well commented

Functionality/Performance: 
- good
- classfication accuracy with the fine-tuned triplet backbone is ~80%

General comments:
- algorithm solves the problem apporpriately with sufficient performance

**Comment by Amelia-Tong** [comment]
> **This is an initial inspection, no action is required at this point**
> 
> Difficulty: Hard
> 
> Readme: Very good
> 
> * Clear flow of information
> * Algortihm and usage explained well
> * Relevant plots for loss and accuracies provided
> * no 2D manifolds are presented
> 
> Commit messages: OK
> 
> Code:
> 
> * tried both siamese and triplet networks
> * good design
> * well commented
> 
> Functionality/Performance:
> 
> * good
> * classfication accuracy with the fine-tuned triplet backbone is ~80%
> 
> General comments:
> 
> * algorithm solves the problem apporpriately with sufficient performance

**Comment by Amelia-Tong** [comment]
Hi @nathasha-naranpanawa ,

Just to confirm that if i need to incorporate manifolds in the readme?

Best regards,
Tong

**Comment by shakes76** [comment]
# Marking
## Good Practice (Design/Commenting, TF/Torch Usage)
Adequate design and implementation
Good spacing and comments
Header blocks missing -1
## Recognition Problem 
Solves problem 
Driver Script present
File structure present
Shows Usage & Demo & Visualisation & Data usage, no UMAPs etc. -2
Module present
Commenting minimal -1
No Data leakage
Difficulty: Hard
## Commit Log
Meaningful commit messages sometimes -1
Progressive commits used
## Documentation
ReadMe acceptable, no refs -1
Model/technical explanation
Good Description and Comments 
Markdown used and PDF submitted
## Pull Request
Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch) 
Feedback required, remove NPY data files -2
Request Description OK, could be better -1

**Comment by shakes76** [comment]
Cannot merge because of data files present, but Feedback marks possible +2 if the requested changes are made (see above).

**Comment by wangzhaomxy** [comment]
No feedback attempt and no feedback marks granted.

**Comment by Amelia-Tong** [comment]
To Incorporating feedback into the pull request, I have:

- added head blockers for all python files;
- added more detailed comments;
- added the reference at the end of the readme file;
- added the UMAPs for original data and embeddings in the readme file;
- removed all data files, including npy files and output files for training processes.
- improved the PR description.

**Comment by wangzhaomxy** [comment]
Feedback attempt and feedback marks granted +2.
